File: leetcode/20.valid-parentheses.ts
/*
 * @lc app=leetcode id=20 lang=typescript
 *
 * [20] Valid Parentheses
 */

// @lc code=start
function isValid(s: string): boolean {
  const openingBrackets = new Set(["(", "{", "["]);
  const closingMap = new Map([
    [")", "("],
    ["}", "{"],
    ["]", "["],
  ]);
  
  const stack: string[] = [];

  for (const c of s) {
    if (openingBrackets.has(c)) {
      stack.push(c);
    } else {
      if (stack.length === 0 || stack[stack.length - 1] !== closingMap.get(c)) {
        return false;
      }
      stack.pop();
    }
  }

  return stack.length === 0;
}
// @lc code=end

File: leetcode/20.valid-parentheses.rs
/*
 * @lc app=leetcode id=20 lang=rust
 *
 * [20] Valid Parentheses
 */

// @lc code=start
use std::collections::HashMap;
impl Solution {
    pub fn is_valid(s: String) -> bool {
        let mut stack = Vec::<char>::new();
        let lookup = HashMap::from(
            [
                (')','('),
                ('}','{'),
                (']','['),
            ]
        );
        for c in s.chars() {
            match c {
                '(' | '{' | '[' => stack.push(c),
                ')' | '}' | ']' => {
                    if *stack.last().unwrap_or(&' ') == lookup[&c] {
                        stack.pop();
                    } else {
                        return false
                    }
                },
                _ => println!("")

            }
        }
        return stack.len() == 0;
    }
}
// @lc code=end

